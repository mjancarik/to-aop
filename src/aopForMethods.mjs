import overOwnProperty from './overOwnProperty.mjs';

export default function aopForMethods(target, pattern) {
  let original = {};
  let prototype = target.prototype;
  while (prototype) {
    overOwnProperty({ target, pattern, original, object: prototype });

    prototype = Reflect.getPrototypeOf(prototype);
  }
}
