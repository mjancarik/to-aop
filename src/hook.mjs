import { AOP_FILTER_FUNCTION } from './symbol.mjs';

export function createHook(name, regular, callback) {
  function hook(meta) {
    return hookFor(meta, regular, callback);
  }

  hook[AOP_FILTER_FUNCTION] = function ({ property, target, prototype }) {
    return (
      (typeof regular === 'string' && property.includes(regular)) ||
      (regular instanceof RegExp && regular.test(property)) ||
      (typeof regular === 'function' &&
        regular({ property, target, name, prototype }))
    );
  };

  return {
    [name]: hook,
  };
}

export function hookFor(meta, regular, callback) {
  if (typeof regular === 'string') {
    if (meta.property.includes(regular)) {
      return callback(meta);
    }

    return null;
  }
  if (regular instanceof RegExp) {
    if (regular.test(meta.property)) {
      return callback(meta);
    }

    return null;
  }

  if (typeof regular === 'function') {
    if (regular(meta)) {
      return callback(meta);
    }

    return null;
  }

  if (Array.isArray(regular)) {
    return regular.map(({ rule, action }) => {
      return hookFor(meta, rule, action || callback);
    });
  }

  throw new TypeError(
    `Invalid rule type ${typeof regular}. Method accept string, regexp, function and array.`,
  );
}

export function hasToRegisterHook(hooks) {
  return (pattern, props) => {
    return hooks.reduce((result, hook) => {
      return (
        result ||
        (pattern[hook] &&
          typeof pattern[hook][AOP_FILTER_FUNCTION] === 'function' &&
          pattern[hook][AOP_FILTER_FUNCTION](props))
      );
    }, false);
  };
}

export const hookName = Object.freeze({
  beforeMethod: 'beforeMethod',
  afterMethod: 'afterMethod',
  aroundMethod: 'aroundMethod',
  beforeGetter: 'beforeGetter',
  afterGetter: 'afterGetter',
  aroundGetter: 'aroundGetter',
  beforeSetter: 'beforeSetter',
  afterSetter: 'afterSetter',
  aroundSetter: 'aroundSetter',
});
