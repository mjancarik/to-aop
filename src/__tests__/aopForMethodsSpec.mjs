import { expect, jest, describe, beforeEach, it } from '@jest/globals';

import aopForMethods from '../aopForMethods.mjs';
import { hookName, createHook } from '../hook.mjs';
import createClasses from './createClasses.mjs';
import createPattern from './createPattern.mjs';

describe('aopForMethods', () => {
  let pattern = null;
  let pattern2 = null;
  let aroundPattern = null;

  describe('class method', () => {
    let beforeMethod = null;
    let afterMethod = null;
    let aroundMethod = null;

    beforeEach(() => {
      beforeMethod = jest.fn();
      afterMethod = jest.fn();
      aroundMethod = jest.fn();

      pattern = createPattern(undefined, {
        [hookName.beforeMethod]: beforeMethod,
        [hookName.afterMethod]: afterMethod,
      });
      pattern2 = createPattern(undefined, {
        [hookName.beforeMethod]: beforeMethod,
        [hookName.afterMethod]: afterMethod,
      });
      aroundPattern = createHook(hookName.aroundMethod, /.*/, aroundMethod);
    });

    it('should call pattern.aroundMethod', () => {
      let { A } = createClasses();

      aopForMethods(A, aroundPattern);
      new A('method').method({}, 1);

      expect(aroundMethod.mock.calls.length).toEqual(1);
      expect(aroundMethod.mock.calls[0][0].original).not.toEqual(aroundMethod);
      expect(aroundMethod.mock.calls[0]).toMatchSnapshot();
    });

    it('should call pattern.aroundMethod three times and modify payload', () => {
      aroundMethod = jest.fn(({ args }) => (args[1] ? args[1] + 1 : args[1]));
      aroundPattern = createHook(hookName.aroundMethod, /.*/, aroundMethod);

      let { A } = createClasses();

      aopForMethods(A, aroundPattern);
      aopForMethods(A, aroundPattern);
      aopForMethods(A, aroundPattern);
      const result = new A('method').method({}, 1);

      expect(result).toEqual(2);
      expect(aroundMethod.mock.calls.length).toEqual(1);
      expect(aroundMethod.mock.calls[0][0].original).not.toEqual(aroundMethod);
      expect(aroundMethod.mock.calls[0]).toMatchSnapshot();
    });

    it('should call pattern.beforeMethod and pattern.afterMethod', () => {
      let { A } = createClasses();

      aopForMethods(A, pattern);
      new A('method').method({}, 1);

      expect(beforeMethod.mock.calls.length).toEqual(1);
      expect(beforeMethod.mock.calls[0]).toMatchSnapshot();
      expect(afterMethod.mock.calls.length).toEqual(1);
      expect(afterMethod.mock.calls[0]).toMatchSnapshot();
    });

    it('should create meta information from pattern.beforeMethod and pass to pattern.afterMethod', () => {
      const value = 'value';
      beforeMethod = ({ meta }) => (meta.information = 'value');
      afterMethod = ({ meta }) => expect(meta.information).toEqual(value);
      pattern = createPattern(undefined, {
        [hookName.beforeMethod]: beforeMethod,
        [hookName.afterMethod]: afterMethod,
      });

      let { A } = createClasses();

      aopForMethods(A, pattern);
      new A('method').method({}, 1);
    });

    it('should call pattern.beforeMethod and pattern.afterMethod for extended class with multiple aspect', () => {
      let { A, C } = createClasses();

      aopForMethods(A, pattern);
      aopForMethods(C, pattern);

      new C('method').method2({}, 1);

      expect(beforeMethod.mock.calls.length).toEqual(1);
      expect(beforeMethod.mock.calls[0]).toMatchSnapshot();
      expect(afterMethod.mock.calls.length).toEqual(1);
      expect(afterMethod.mock.calls[0]).toMatchSnapshot();
    });

    it('should call pattern.arroundMethod for extended class with multiple aspect', () => {
      let { A, C } = createClasses();

      aopForMethods(A, aroundPattern);
      aopForMethods(C, aroundPattern);
      new C('method').method2({}, 1);

      expect(aroundMethod.mock.calls.length).toEqual(1);
      expect(aroundMethod.mock.calls[0][0].original).not.toEqual(aroundMethod);
      expect(aroundMethod.mock.calls[0]).toMatchSnapshot();
    });

    it('should call pattern.beforeMethod and pattern.afterMethod for extended class', () => {
      let { B } = createClasses();

      aopForMethods(B, pattern);
      new B('method').method({}, 1);

      expect(beforeMethod.mock.calls.length).toEqual(1);
      expect(beforeMethod.mock.calls[0]).toMatchSnapshot();
      expect(afterMethod.mock.calls.length).toEqual(1);
      expect(afterMethod.mock.calls[0]).toMatchSnapshot();
    });

    it('should call pattern.beforeMethod and pattern.afterMethod after class is instanced', () => {
      let { B } = createClasses();

      const b = new B('method');
      aopForMethods(B, pattern);
      b.method({}, 1);

      expect(beforeMethod.mock.calls.length).toEqual(1);
      expect(beforeMethod.mock.calls[0]).toMatchSnapshot();
      expect(afterMethod.mock.calls.length).toEqual(1);
      expect(afterMethod.mock.calls[0]).toMatchSnapshot();
    });

    it('should create meta information from pattern.beforeMethod and pass to pattern.afterMethod after class is instanced', () => {
      const value = 'value';
      beforeMethod = ({ meta }) => (meta.information = 'value');
      afterMethod = ({ meta }) => expect(meta.information).toEqual(value);
      pattern = createPattern(undefined, {
        [hookName.beforeMethod]: beforeMethod,
        [hookName.afterMethod]: afterMethod,
      });

      let { B } = createClasses();
      const b = new B('method');

      aopForMethods(B, pattern);
      b.method({}, 1);
    });

    it('should call pattern.beforeMethod and pattern.afterMethod for extended classes with same ancestor', () => {
      let { B, C, A } = createClasses();

      const b = new B('method');
      aopForMethods(A, pattern);
      aopForMethods(A, pattern2);
      aopForMethods(C, pattern);
      aopForMethods(C, pattern2);
      aopForMethods(B, pattern);
      aopForMethods(B, pattern2);

      expect(b.method({}, 1)).toEqual('B method');

      expect(beforeMethod.mock.calls.length).toEqual(6);
      expect(beforeMethod.mock.calls[0]).toMatchSnapshot();
      expect(afterMethod.mock.calls.length).toEqual(6);
      expect(afterMethod.mock.calls[0]).toMatchSnapshot();
    });

    it('should call pattern.beforeMethod and pattern.afterMethod for super.superMethod', () => {
      let { C } = createClasses();
      pattern = createPattern(
        {
          [hookName.beforeMethod]: 'superMethod',
          [hookName.afterMethod]: 'superMethod',
        },
        {
          [hookName.beforeMethod]: beforeMethod,
          [hookName.afterMethod]: afterMethod,
        },
      );

      aopForMethods(C, pattern);
      const c = new C('super method');

      expect(c.superMethod()).toEqual('super method');

      expect(beforeMethod.mock.calls.length).toEqual(1);
      expect(beforeMethod.mock.calls[0]).toMatchSnapshot();
      expect(afterMethod.mock.calls.length).toEqual(1);
      expect(afterMethod.mock.calls[0]).toMatchSnapshot();
    });

    it('should call pattern.aroundMethod for super.superMethod', () => {
      let { C } = createClasses();

      aopForMethods(C, aroundPattern);
      const c = new C('super method');

      expect(c.superMethod()).toEqual(undefined);

      expect(aroundMethod.mock.calls.length).toEqual(1);
      expect(aroundMethod.mock.calls[0][0].original).not.toEqual(aroundMethod);
      expect(aroundMethod.mock.calls[0]).toMatchSnapshot();
    });
  });

  describe('es5 problems', () => {
    let beforeMethod = null;
    let afterMethod = null;

    beforeEach(() => {
      beforeMethod = jest.fn();
      afterMethod = jest.fn();

      pattern = createPattern(undefined, {
        [hookName.beforeMethod]: beforeMethod,
        [hookName.afterMethod]: afterMethod,
      });
    });

    it('should be same instance', () => {
      let { B, A } = createClasses();

      const b = new B('method');
      aopForMethods(B, pattern);
      b.method({}, 1);

      expect(beforeMethod.mock.calls.length).toEqual(1);
      expect(afterMethod.mock.calls.length).toEqual(1);
      expect(b instanceof A).toBeTruthy();
      expect(b instanceof B).toBeTruthy();
    });
  });
});
