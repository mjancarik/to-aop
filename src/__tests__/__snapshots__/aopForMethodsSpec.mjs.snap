// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`aopForMethods class method should call pattern.aroundMethod 1`] = `
[
  {
    "args": [
      {},
      1,
    ],
    "context": A {
      "variable": "method",
    },
    "meta": {},
    "object": {
      "_privateMethod": [Function],
      "method": [Function],
      "superMethod": [Function],
    },
    "original": [Function],
    "property": "method",
    "target": [Function],
  },
]
`;

exports[`aopForMethods class method should call pattern.aroundMethod for super.superMethod 1`] = `
[
  {
    "args": [],
    "context": C {
      "map": Map {},
      "variable": "super method",
    },
    "meta": {},
    "object": {
      "_privateMethod": [Function],
      "method": [Function],
      "method2": [Function],
      "superMethod": [Function],
    },
    "original": [Function],
    "property": "superMethod",
    "target": [Function],
  },
]
`;

exports[`aopForMethods class method should call pattern.aroundMethod three times and modify payload 1`] = `
[
  {
    "args": [
      {},
      1,
    ],
    "context": A {
      "variable": "method",
    },
    "meta": {},
    "object": {
      "_privateMethod": [Function],
      "method": [Function],
      "superMethod": [Function],
    },
    "original": [Function],
    "property": "method",
    "target": [Function],
  },
]
`;

exports[`aopForMethods class method should call pattern.arroundMethod for extended class with multiple aspect 1`] = `
[
  {
    "args": [
      {},
      1,
    ],
    "context": C {
      "map": Map {},
      "variable": "method",
    },
    "meta": {},
    "object": {
      "_privateMethod": [Function],
      "method": [Function],
      "method2": [Function],
      "superMethod": [Function],
    },
    "original": [Function],
    "property": "method2",
    "target": [Function],
  },
]
`;

exports[`aopForMethods class method should call pattern.beforeMethod and pattern.afterMethod 1`] = `
[
  {
    "args": [
      {},
      1,
    ],
    "context": A {
      "variable": "method",
    },
    "meta": {},
    "object": {
      "_privateMethod": [Function],
      "method": [Function],
      "superMethod": [Function],
    },
    "property": "method",
    "target": [Function],
  },
]
`;

exports[`aopForMethods class method should call pattern.beforeMethod and pattern.afterMethod 2`] = `
[
  {
    "args": [
      {},
      1,
    ],
    "context": A {
      "variable": "method",
    },
    "meta": {},
    "object": {
      "_privateMethod": [Function],
      "method": [Function],
      "superMethod": [Function],
    },
    "payload": "method",
    "property": "method",
    "target": [Function],
  },
]
`;

exports[`aopForMethods class method should call pattern.beforeMethod and pattern.afterMethod after class is instanced 1`] = `
[
  {
    "args": [
      {},
      1,
    ],
    "context": B {
      "variable": "method",
    },
    "meta": {},
    "object": {
      "_privateMethod": [Function],
      "method": [Function],
      "superMethod": [Function],
    },
    "property": "method",
    "target": [Function],
  },
]
`;

exports[`aopForMethods class method should call pattern.beforeMethod and pattern.afterMethod after class is instanced 2`] = `
[
  {
    "args": [
      {},
      1,
    ],
    "context": B {
      "variable": "method",
    },
    "meta": {},
    "object": {
      "_privateMethod": [Function],
      "method": [Function],
      "superMethod": [Function],
    },
    "payload": "B method",
    "property": "method",
    "target": [Function],
  },
]
`;

exports[`aopForMethods class method should call pattern.beforeMethod and pattern.afterMethod for extended class 1`] = `
[
  {
    "args": [
      {},
      1,
    ],
    "context": B {
      "variable": "method",
    },
    "meta": {},
    "object": {
      "_privateMethod": [Function],
      "method": [Function],
      "superMethod": [Function],
    },
    "property": "method",
    "target": [Function],
  },
]
`;

exports[`aopForMethods class method should call pattern.beforeMethod and pattern.afterMethod for extended class 2`] = `
[
  {
    "args": [
      {},
      1,
    ],
    "context": B {
      "variable": "method",
    },
    "meta": {},
    "object": {
      "_privateMethod": [Function],
      "method": [Function],
      "superMethod": [Function],
    },
    "payload": "B method",
    "property": "method",
    "target": [Function],
  },
]
`;

exports[`aopForMethods class method should call pattern.beforeMethod and pattern.afterMethod for extended class with multiple aspect 1`] = `
[
  {
    "args": [
      {},
      1,
    ],
    "context": C {
      "map": Map {},
      "variable": "method",
    },
    "meta": {},
    "object": {
      "_privateMethod": [Function],
      "method": [Function],
      "method2": [Function],
      "superMethod": [Function],
    },
    "property": "method2",
    "target": [Function],
  },
]
`;

exports[`aopForMethods class method should call pattern.beforeMethod and pattern.afterMethod for extended class with multiple aspect 2`] = `
[
  {
    "args": [
      {},
      1,
    ],
    "context": C {
      "map": Map {},
      "variable": "method",
    },
    "meta": {},
    "object": {
      "_privateMethod": [Function],
      "method": [Function],
      "method2": [Function],
      "superMethod": [Function],
    },
    "payload": Map {},
    "property": "method2",
    "target": [Function],
  },
]
`;

exports[`aopForMethods class method should call pattern.beforeMethod and pattern.afterMethod for extended classes with same ancestor 1`] = `
[
  {
    "args": [
      {},
      1,
    ],
    "context": B {
      "variable": "method",
    },
    "meta": {},
    "object": {
      "_privateMethod": [Function],
      "method": [Function],
      "superMethod": [Function],
    },
    "property": "method",
    "target": [Function],
  },
]
`;

exports[`aopForMethods class method should call pattern.beforeMethod and pattern.afterMethod for extended classes with same ancestor 2`] = `
[
  {
    "args": [
      {},
      1,
    ],
    "context": B {
      "variable": "method",
    },
    "meta": {},
    "object": {
      "_privateMethod": [Function],
      "method": [Function],
      "superMethod": [Function],
    },
    "payload": "method",
    "property": "method",
    "target": [Function],
  },
]
`;

exports[`aopForMethods class method should call pattern.beforeMethod and pattern.afterMethod for super.superMethod 1`] = `
[
  {
    "args": [],
    "context": C {
      "map": Map {},
      "variable": "super method",
    },
    "meta": {},
    "object": {
      "__defineGetter__": [Function],
      "__defineSetter__": [Function],
      "__lookupGetter__": [Function],
      "__lookupSetter__": [Function],
      "_privateMethod": [Function],
      "constructor": [Function],
      "hasOwnProperty": [Function],
      "isPrototypeOf": [Function],
      "method": [Function],
      "method2": [Function],
      "propertyIsEnumerable": [Function],
      "superMethod": [Function],
      "toLocaleString": [Function],
      "toString": [Function],
      "valueOf": [Function],
    },
    "property": "superMethod",
    "target": [Function],
  },
]
`;

exports[`aopForMethods class method should call pattern.beforeMethod and pattern.afterMethod for super.superMethod 2`] = `
[
  {
    "args": [],
    "context": C {
      "map": Map {},
      "variable": "super method",
    },
    "meta": {},
    "object": {
      "__defineGetter__": [Function],
      "__defineSetter__": [Function],
      "__lookupGetter__": [Function],
      "__lookupSetter__": [Function],
      "_privateMethod": [Function],
      "constructor": [Function],
      "hasOwnProperty": [Function],
      "isPrototypeOf": [Function],
      "method": [Function],
      "method2": [Function],
      "propertyIsEnumerable": [Function],
      "superMethod": [Function],
      "toLocaleString": [Function],
      "toString": [Function],
      "valueOf": [Function],
    },
    "payload": "super method",
    "property": "superMethod",
    "target": [Function],
  },
]
`;
